### PR TITLE
Add LightningAddress class

### PR DESCRIPTION
**Pull request by bumi**
to load info for a lightning address

this can be used as inspiration: https://github.com/dolcalmi/lnurl-pay
but I don't like that it has that much dependencies. 

ref #1

### CODE REVIEW

File: src/lightning-address.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  domain: string | undefined;
+  keysendData: unknown;
+  lnurlpData: unknown;
+
+  constructor(address: string, options: { proxy: boolean }) {
+    this.address = address;
+    this.options = { proxy: true, ...options };
+    this.parse();
+    if (this.options.proxy) {
+      this.fetchWithCorsProxy();
```
Comment (reneaaron) [inline src/lightning-address.ts:19]: This shouldn't be done in the constructor.

**Comment by bumi** [comment]
I am still experimenting with the API and how it feels right to use it.

**Comment by reneaaron** [comment]
Is this ready to merge? To me that is the most used thing in this toolset and I would like to play around with it for a bit.

File: src/lightning-address.ts [hunk 2]
```diff
@@ -64,6 +67,41 @@ export default class LightningAddress {
...
+      `The comment length must be ${commentAllowed} characters or fewer`
+    )
+    if (comment) invoiceParams.comment = comment
+
+    callbackUrl.search = new URLSearchParams(invoiceParams).toString()
+  
+    const data = await fetch(callbackUrl)
+    const json = await data.json()
+    const paymentRequest = json && json.pr && json.pr.toString()
+    const verifyURL = json && (json.verify || '').toString()
```
Comment (bumi) [inline src/lightning-address.ts:96]: does this need to be a sting or can it be undefined/null?

File: src/invoice.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+export default class Invoice {
+  paymentRequest: string;
+  paymentHash: string;
+  verify: string;
+
+  constructor(lnUrlPayResponse: {pr: string, verify: string}) {
+    this.paymentRequest = lnUrlPayResponse.pr;
+    this.verify = lnUrlPayResponse.verify;
+    this.paymentHash = getHashFromInvoice(this.paymentRequest) as string;
```
Comment (bumi) [inline src/invoice.ts:14]: let's also allow to pass in the preimage here. 

then we can do: 

```
const invoice = new Invoice(params);

if (invoice.isPaid()) {
  ...
}
```

the `isPaid()` function then either checks: 
1. preimage with the payment hash
2. the verify URL
3. thows an error if both does not work

File: src/lightning-address.ts (hunk 2 above)
Comment (im-adithya) [reply src/lightning-address.ts:96]: it can be undefined or null

File: src/invoice.ts (hunk 3 above)
Comment (im-adithya) [reply src/invoice.ts:15]: Done!

File: src/lightning-address.ts (hunk 2 above)
Comment (bumi) [reply src/lightning-address.ts:96]: yes, then we write an empty string in `verifyURL` ? is that a good practice?

File: src/lightning-address.ts [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+    if (comment && commentAllowed > 0 && comment.length > commentAllowed)
+    throw new Error(
+      `The comment length must be ${commentAllowed} characters or fewer`
+    )
+    if (comment) invoiceParams.comment = comment
+
+    callbackUrl.search = new URLSearchParams(invoiceParams).toString()
+  
+    const data = await fetch(callbackUrl)
```
Comment (bumi) [inline src/lightning-address.ts:94]: here we probably also have to use the proxy. 

similar to `fetch()` - if the proxy should be used do the request through the proxy. 

so `if (this.options.proxy) {` 

and for that we need a new endpoint in the proxy app: https://github.com/getAlby/lightning-address-details-proxy/blob/master/main.go

that endpoint accepts a lightning address and the LNURL params. It then fetches the callback URL and does a request to that callback URL with all other params and simply returns the response.

File: src/lightning-address.ts [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    const paymentRequest = json && json.pr && json.pr.toString()
+    if (!paymentRequest) throw new Error('Invalid pay service invoice')
+
+    const invoiceArgs: InvoiceArgs = {pr: paymentRequest};
+    if (json && json.verify) invoiceArgs.verify = json.verify.toString();
+
+    return new Invoice(invoiceArgs);
+  }
+
+  async requestInvoice(
```
Comment (bumi) [inline src/lightning-address.ts:85]: should this method also be used by the zap() call? 

this validates the amounts for example (which is currently not done in the zap() function)

File: src/invoice.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    } else {
+      throw new Error('Could not verify payment');
+    }
+  }
+
+  validatePreimage(preimage: string): boolean {
+    if (!preimage || !this.paymentHash) return false
+  
+    try {
+      const preimageHash = sha256(preimage).toString(Hex)
```
Comment (bumi) [inline src/invoice.ts:34]: we need to parse the preimage as hex and then do the sha256

**Pull request merged**